lib/screens/v3_api/parser.ex

```diff
@@ -0,0 +1,109 @@
+defmodule Screens.V3Api.Parser do
+  @moduledoc "Tools for parsing V3 API responses."
+
+  @parsers %{
+    "alert" => Screens.Alerts.Parser,
+    "facility" => Screens.Facilities.Parser,
+    "prediction" => Screens.Predictions.Parser,
+    "route" => Screens.Routes.Parser,
+    "route_pattern" => Screens.RoutePatterns.Parser,
+    "schedule" => Screens.Schedules.Parser,
+    "stop" => Screens.Stops.Parser,
+    "trip" => Screens.Trips.Parser,
+    "vehicle" => Screens.Vehicles.Parser
+  }
+
+  @typep included :: %{{id :: String.t(), type :: String.t()} => object()}
+  @typep object :: %{String.t() => term()}
+
+  @doc "Parse a complete JSON-decoded V3 API response."
+  @spec parse(response :: object()) :: struct() | [struct()]
+  def parse(%{"data" => data} = response) when is_map(data) do
+    [resource] = parse(%{response | "data" => [data]})
+    resource
+  end
+
+  def parse(%{"data" => data} = response) when is_list(data) do
+    included =
+      response
+      |> Map.get("included", [])
+      |> then(&Enum.concat(data, &1))
+      |> Map.new(fn %{"id" => id, "type" => type} = resource -> {{id, type}, resource} end)
+
+    Enum.map(data, &parse_resource(&1, included))
+  end
+
+  @doc """
+  Parse a single resource. Expects as the second argument a map of all resources that were present
+  in the response, keyed by ID and type, for resolving relationship references.
+  """
+  @spec parse_resource(object(), included()) :: struct()
+  def parse_resource(
+        %{
+          "type" => "line",
+          "id" => id,
+          "attributes" => %{
+            "long_name" => long_name,
+            "short_name" => short_name,
+            "sort_order" => sort_order
+          }
+        },
+        _included
+      ) do
+    %Screens.Lines.Line{
+      id: id,
+      long_name: long_name,
+      short_name: short_name,
+      sort_order: sort_order
+    }
+  end
+
+  for {type, module} <- @parsers do
+    def parse_resource(%{"type" => unquote(type)} = data, included),
+      do: unquote(module).parse(data, included)
+  end
```

Comment:
I love the use of `unquote` here! Very cool.

Is there are a reason that `Line` parsing needs to be defined separately instead of defined in its own module like the other resource types? It would make more sense to me to have that defined like every other included resource

Reply:
It doesn't need to be, I just thought the logic was trivial enough that there was no need for it to live in a separate module. Originally I thought maybe that could be true of other resources (that their logic was minimal enough that it could also live here) but it turned out Line was the only "simple" one. I can put that back in its own module as well.

Reply:
Gotcha, I think putting that back in it's own module would make it much cleaner in that case